riscv-runtime/src/io.rs

```diff
@@ -0,0 +1,70 @@
+use core::arch::asm;
+
+extern crate alloc;
+
+use powdr_riscv_syscalls::Syscall;
+
+use alloc::vec;
+use alloc::vec::Vec;
+
+/// Reads a single u32 from the file descriptor fd.
+pub fn read_u32(fd: u32) -> u32 {
+    let mut value: u32;
+    unsafe {
+        asm!("ecall", lateout("a0") value, in("a0") fd, in("t0") u32::from(Syscall::Input));
+    }
+    value
+}
+
+/// Reads data.len() u32s from the file descriptor fd into the data slice.
+pub fn read_slice(fd: u32, data: &mut [u32]) {
+    for (i, d) in data.iter_mut().enumerate() {
+        unsafe {
+            asm!("ecall", lateout("a0") *d, in("a0") fd, in("a1") (i+1) as u32, in("t0") u32::from(Syscall::DataIdentifier))
+        };
+    }
+}
+
+/// Reads the length of the data first at index 0, then the data itself.
+pub fn read_data_len(fd: u32) -> usize {
+    let mut out: u32;
+    unsafe {
+        asm!("ecall", lateout("a0") out, in("a0") fd, in("a1") 0, in("t0") u32::from(Syscall::DataIdentifier))
+    };
+    out as usize
+}
+
+/// Writes a single u32 to the file descriptor fd.
+pub fn write_u8(fd: u32, w: u8) {
```

Comment:
very minor, but `w` was from "word" right?

Reply:
yep, changed it to `byte`